docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,424 @@
+
+The `kubelet` will error if the defined `--system-container` does not exist
+on `systemd` environments.  It will verify that the appropriate `cpu` and `memory`
+controllers are enabled.
+
+### Kubernetes reserved
+
+The node may set aside a set of resources for Kubernetes components:
+
+* `--kube-reserved=:` - A set of `ResourceName`=`ResourceQuantity` pairs that
+describe resources reserved for host daemons.
+
+The `kubelet` does not enforce `--kube-reserved` at this time, but the ability
+to distinguish the static reservation from observed usage is important for node accounting.
+
+This proposal asserts that `kubernetes.slice` is the default slice associated with
+the `kubelet` and `kube-proxy` service units defined in the project.  Keeping it
+separate from `system.slice` allows for accounting to be distinguished separately.
+
+The `kubelet` will detect its `cgroup` to track `kube-reserved` observed usage.  If
+the `kubelet` detects that its a child of the `system-container` based on the observed
+`cgroup` hierarchy, it will warn.
+
+If the `kubelet` is launched directly from a terminal, it's most likely destination will
+be in a `scope` that is a child of `user.slice` as follows:
+
+`/sys/fs/cgroup/<controller>/user.slice/user-1000.slice/session-1.scope`
+
+In this context, the parent `scope` is what will be used to facilitate local developer
+debugging scenarios for tracking `kube-reserved` usage.
+
+### Kubernetes container runtime reserved
+
+This proposal asserts that the reservation of compute resources for any associated
+container runtime daemons is tracked by the operator under the `system-reserved` or
+`kubernetes-reserved` values and any enforced limits are set by the
+operator specific to the container runtime.
+
+**Docker**
+
+If the `kubelet` is configured with the `container-runtime` set to `docker`, the
+`kubelet` will detect the `cgroup` associated with the `docker` daemon and use that
+to do local node accounting.  If an operator wants to impose runtime limits on the
+`docker` daemon to control resource usage, the operator should set those explicitly in
+the `service` unit that launches `docker`.  The `kubelet` will not set any limits itself
+at this time and will assume whatever budget was set aside for `docker` was included in
+either `--kube-reserved` or `--system-reserved` reservations.
+
+Many OS distributions package `docker` by default, and it will often belong to the
+`system.slice` hierarchy, and therefore operators will need to budget it for there
+by default unless they explicitly move it.
+
+**rkt**
+
+rkt has no client/server daemon, and therefore has no explicit requirements on container-runtime
+reservation.
+
+### kubelet cgroup enforcement
+
+The `kubelet` does not enforce the `system-reserved` or `kube-reserved` values by default.
+
+The `kubelet` should support an additional flag to turn on enforcement:
+
+* `--cgroup-enforce=false` - Optional flag that if true tells the `kubelet` to enforce
```

Comment:
I believe the original intention was for `kube-reserved` to be enforced, but not `system-reserved`. It might be useful to make this flag a `[]string` which could take values `{<empty>|kube|system|kube,system}`, or is that over-generalizing?